### ISSUE TITLE
Use Teal as a typed Lua alternative

### ISSUE THREAD

**Issue by britzl** [defold/defold#8284]
Follow-up prototyping task to https://github.com/defold/defold/issues/6398

**Comment by u546342** [comment]
Hello. It's great that you are exploring options for adding type annotations to Defold.

I would like to add my 5 cents, as this topic was actively discussed in the Russian-speaking Defold chat.

When we were exploring the possibility of integrating Teal into our workflow, the main limitation that stood out for us was the restriction on union types in Teal: [Teal Documentation on Current Limitations of Union Types](https://github.com/teal-language/tl/blob/master/docs/tutorial.md#current-limitations-of-union-types)

This made it challenging to create definitions even for the standard Defold API in certain situations. The simplest example is the function `vmath.length`, which accepts a union type of `vector3|vector4|quat`.

This limitation led us to switch to EmmyLua and later to sumneko's Lua Language Server, proved to be much more functional, albeit with a lot of problems of its own.

**Comment by dri-richard** [comment]
How would Teal affect message passing? It's pretty common to send structured data in a message, and because it's more loosely-coupled than a function call it can be harder to debug. It's convention to add a comment above a function listing the parameters and their types, but it's not so common for messages.
So type checking on the send & receive would be useful.

**Comment by vlaaad** [comment]
I implemented the initial support in Editor and Bob. This will be available for evaluation in the next beta (with https://github.com/defold/extension-teal)

**Comment by astrochili** [comment]
@vlaaad Hi, where can I find information on how to use Teal with Bob? Or installing the extension is enougth?

**Comment by vlaaad** [comment]
I don't think you can use it straight away, not until a beta release at least. It just got merged into the `dev` branch. But yes, once beta release, you'll only need an extension to try it out
